### DISCUSSION TITLE
Is there any plugin or recipe for a SAML login?

### DISCUSSION THREAD

**Discussion by stfnlutter** [cypress-io/cypress#19173] [Questions and Help] [answered]
Hello, I am new to Cypress and one of my first tasks is to get beyond the SAML login of our application. Of course going through the UI works but is a bit flaky, and I'd prefer the recommended way of doing this without UI. I've did a bit of research to see if there is anything which would help me with that, but the [cypress-example-recipes](https://github.com/cypress-io/cypress-example-recipes) or [authentication plugins](https://docs.cypress.io/plugins/index#authentication) don't seem to provide anything in this regards nor do I find anything else on the web besides the occasional person struggling with the same thing.

Have I overlooked anything or is this indeed something I need to implement manually? Just prefer to double check before I invest myself in this as this doesn't seem trivial... :)

**Comment by portyu9** [comment, marked as answer]
Hi @stfnlutter 

There isn't a generic Cypress plugin that can programmatically authenticate every SAML implementation because the SAML/IdP flow is provider-specific.

The recommended solution is to bypass the login UI and establish the authenticated application session directly—for example through an authentication/test API—and then cache it with `cy.session()`:

```js
cy.session('user', () => {
  cy.request('POST', '/test/login', {
    username: Cypress.env('username'),
  })
})
```

If your IdP exposes an API, use that to obtain the session/token instead.

If the SAML UI flow cannot be bypassed, use `cy.origin()` for the IdP domain and wrap the login in `cy.session()`.

So there is no SAML-specific Cypress plugin you are missing—the programmatic implementation has to match your application's authentication system.

Please mark this answer as accepted if it helped.

**Reply by portyu9** [reply]
@stfnlutter Please mark this answer as accepted if it helped, thank you.

**Reply by stfnlutter** [reply]
Thank you very much for your answer! I will close this issue
